File: source_code/C++/binary_search/binary_search.cpp
#include <iostream>
using namespace std;

int binary_search(int target, int * data, int data_length)
{
	int start = 0;
	int end = data_length - 1;

	while (start <= end)
	{
		int mid = (start + end) / 2;

		if (data[mid] == target)
			return mid;
		else if (data[mid] > target)
			end = mid - 1;
		else
			start = mid + 1;
	}

	return -1;
}

int main(void)
{
	int arr[] = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
	
	int target = 4;
	int length = sizeof(arr) / sizeof(int);
	int idx = binary_search(target, arr, length);
	if (idx == -1)
		cout << "찾는 데이터가 존재하지 않습니다." << endl;
	else
	{
		cout << "index : " << idx << endl;
		cout << "data : " << arr[idx] << endl;
	}

	target = 15;
	idx = binary_search(target, arr, length);
	if (idx == -1)
		cout << "찾는 데이터가 존재하지 않습니다." << endl;
	else
	{
		cout << "index : " << idx << endl;
		cout << "data : " << arr[idx] << endl;
	}


	return 0;
}

File: source_code/python/datastructure/binary_search/binary_search.py
def binary_search(target, data):
    data.sort()
    start = 0
    end = len(data) - 1

    while start <= end:
        mid = (start + end)//2

        if data[mid] == target:
            return mid
        elif data[mid] > target:
            end = mid - 1
        else:
            start = mid + 1

    return -1

if __name__ == "__main__":
    data = [i**2 for i in range(1, 11)]

    target = 9
    idx = binary_search(target, data)

    if idx == -1:
        print("{}이 존재하지 않습니다".format(target))
    else:
        print("찾는 데이터의 인덱스는 {}이고 데이터는 {} 입니다".format(idx, data[idx]))

    
